File: benchmark/python/modular_exp.py
import time
start = time.time() * 1000
def mod_exp(base, exp, mod):
    result = 1
    b = base % mod
    e = exp
    while e > 0:
        if e % 2 == 1:
            result = (result * b) % mod
        b = (b * b) % mod
        e = e // 2
    return result

s = 0
for _ in range(10000):
    s += mod_exp(3, 1000, 1000000007)
end = time.time() * 1000
print(f"Sum: {s}")
print(f"Time: {end - start:.0f}ms")

File: benchmark/php/modular_exp.php
<?php
function mod_exp($base, $exp, $mod) {
    $result = 1; $base = $base % $mod;
    while ($exp > 0) {
        if ($exp % 2 == 1) $result = ($result * $base) % $mod;
        $exp = intdiv($exp, 2); $base = ($base * $base) % $mod;
    }
    return $result;
}
$start = hrtime(true);
$sum = 0;
for ($i = 0; $i < 10000; $i++) { $sum += mod_exp(3, 1000, 1000000007); }
$end = hrtime(true);
$ms = ($end - $start) / 1e6;
echo "Sum: " . $sum . "
";
echo "Time: " . round($ms) . "ms
";
